app/src/sharedTest/java/org/oppia/android/app/profile/ProfileChooserFragmentTest.kt

```diff
@@ -454,7 +466,7 @@ class ProfileChooserFragmentTest {
       testCoroutineDispatchers.runCurrent()
       onView(withId(R.id.administrator_controls_linear_layout)).perform(click())
       intended(hasComponent(AdminAuthActivity::class.java.name))
-      intended(hasExtra(ADMIN_AUTH_ENUM_EXTRA_KEY, AdminAuthEnum.PROFILE_ADMIN_CONTROLS.value))
+      intended(hasExtraWithKey(ADMIN_AUTH_ACTIVITY_ARGUMENTS_KEY))
```

Comment:
@adhiamboperes, could you please take a look at this code and let me know if it's fine?

Reply:
This depends.
My thoughts is to use hasProtoextra here instead of `hasExtraWithKey`.

Remember, in this PR we are introducing protos, so the test for every actiivity should verify that the correct proto is being passed.

For the activities that introduce decorateWithCurrentProfileId, we will add an extra check for `intended(hasExtraWithKey(PROFILE_ID_INTENT_DECORATOR)`.

This way, we test both new things introduced in this PR.

Reply:
If you search the method like this:
![image](https://github.com/oppia/oppia-android/assets/59600948/d0f7cc1a-d99b-4471-afe1-a9892fda8232)

You will see how it is used in the existing tests, and do the same for your changes.

Reply:
> This depends. My thoughts is to use hasProtoextra here instead of `hasExtraWithKey`.
> 
> Remember, in this PR we are introducing protos, so the test for every actiivity should verify that the correct proto is being passed.
> 
> For the activities that introduce decorateWithCurrentProfileId, we will add an extra check for `intended(hasExtraWithKey(PROFILE_ID_INTENT_DECORATOR)`.
> 
> This way, we test both new things introduced in this PR.

I did the same u are saying, but  test failed because of passing Single argument( in this test it is enum value ). So I had not option other than using hasExtraKey().
I will look into it more, will update you later.

Reply:
I don't understand the issue you are facing. Can you show me the code that you have written, so that I can advise?

Reply:
> I don't understand the issue you are facing. Can you show me the code that you have written, so that I can advise?

```
val args = AdminAuthActivityArguments.newBuilder().apply {
        this.adminPinEnum = AdminAuthEnum.PROFILE_ADMIN_CONTROLS.value
      }.build()
      intended(hasComponent(AdminAuthActivity::class.java.name))
      intended(hasProtoExtra(ADMIN_AUTH_ACTIVITY_ARGUMENTS_KEY,args))
```
Here only passed adminPinEnum.

Reply:
> > I don't understand the issue you are facing. Can you show me the code that you have written, so that I can advise?
> 
> ```
> val args = AdminAuthActivityArguments.newBuilder().apply {
>         this.adminPinEnum = AdminAuthEnum.PROFILE_ADMIN_CONTROLS.value
>       }.build()
>       intended(hasComponent(AdminAuthActivity::class.java.name))
>       intended(hasProtoExtra(ADMIN_AUTH_ACTIVITY_ARGUMENTS_KEY,args))
> ```
> 
> Here only passed adminPinEnum, should i pass all arguments ? if yes what should be the value ?.

Hi @adhiamboperes, PTAL at code.

Reply:
Yes, you must pass the arguments. The arguments need to correspond to the ones required by the class under test. You will create the arguments as part of the test setup.


Here's a code example: https://github.com/oppia/oppia-android/blob/e81ba7192d021101d30e686b752dae97f1f84cbc/app/src/sharedTest/java/org/oppia/android/app/home/RecentlyPlayedFragmentTest.kt#L845

Reply:
```
   val args = AdminAuthActivityArguments.newBuilder().apply {
        this.adminPin = "12345"
        this.internalProfileId = 0
        this.colorRgb = -1234
        this.adminPinEnum = AdminAuthEnum.PROFILE_ADMIN_CONTROLS.value
      }.build()
      intended(hasComponent(AdminAuthActivity::class.java.name))
      intended(hasProtoExtra(ADMIN_AUTH_ACTIVITY_ARGUMENTS_KEY,args))
```
Yes, that's right but what would be the values for those proto fields, i just tried with random values so test fails.

Reply:
In the test file, I have seen this function for creating the intent in the tests. These same values are what we consider `expectedParams` and we  need to verify that they exist.

<img width="1080" alt="Screenshot 2023-12-21 at 00 10 16" src="https://github.com/oppia/oppia-android/assets/59600948/87f17316-204f-4c4b-acf4-bc80de11cff5">

Reply:
```
  @Test
  fun testProfileChooserFragment_clickAddProfile_opensAdminAuthActivity() {
    profileTestHelper.initializeProfiles(autoLogIn = false)
    launch<ProfileChooserActivity>(createProfileChooserActivityIntent()).use {
      testCoroutineDispatchers.runCurrent()
      onView(
        atPosition(
          recyclerViewId = R.id.profile_recycler_view,
          position = 4
        )
      ).perform(click())
      intended(hasComponent(AdminAuthActivity::class.java.name))
    }
  }
```
In this test we are creating intent for **ProfileChooserActivity**, so i don't see any values for AdminAuthActivity to pass.

Reply:
I see this is a different activity but I don't understand your confusion.
1. Each activity has an intent that is used to create it. 
2.  In every test, that intent is called using a method that is called either `openXyzActivity()` or `createXyzActivtyIntent()` 
3. When you navigate to these methods, you will see the arguments that gave been passed to that intent.
4. In our case, there can be an extra(string, proto, etc).
5. Sometimes, we also decorate the intent with profile ID.
6. Sometimes, like with the profile chooser, there is no bundle extra passed to the intent. So we only have the assertion for component, which is mandatory for all intents. Nothing else.
private fun createProfileChooserActivityIntent(): Intent {
    return ProfileChooserActivity
      .createProfileChooserActivity(ApplicationProvider.getApplicationContext())
  }
7. When testing the intents, you need to test for what is there. All intents must have a component to hich is the Class to be opened, so all tests must have the hasComponent() assertion.
8. Some activities will not have other exrras but will have been decorated with profileId, so in addition to hasComponent, there will be an assertion to check if intended has a profileId (ee discussed how to do this in a different thread).
9. Some activities have only a proto extra in the bundle, so your test must contain an hasProtoextra() assertion that expects an argument, and you will create this as per the example in my last comment.
10. Other times, an intent will have both the extra args, and will also have decorate with profile Id. We expect to see 3 assertions for component, protoextra and profileId.
11. It is your responsibility to countercheck that the intent in the test file matches the corresponding production class intent. You must make note of the items that are in the intent, and assert them in your test.
12. Most of the tests already have the correct assertions, and the major changes in them will be to replace hasextra, hasextrawith key with hasProtoextra. Or add the check for profileId. Or both.


Does this help make things clearer?

Reply:
Referencing 9th point
I'm asking about the values for the arguments which I assert using hasProtoaExtra() for adminauthacitvity, I'm not getting what values to give because there is not intent creation for Adminauthacitvity in this test file.

In this test file only profilechooserfragment intent creation is there.

Reply:
> I see this is a different activity but I don't understand your confusion.
> 
> 1. Each activity has an intent that is used to create it.
> 2. In every test, that intent is called using a method that is called either `openXyzActivity()` or `createXyzActivtyIntent()`
> 3. When you navigate to these methods, you will see the arguments that have been passed to that intent.
> 4. In our case, there can be an extra(string, proto, etc).
> 5. Sometimes, we also decorate the intent with profile ID.
> 6. Sometimes, like with the profile chooser, there is no bundle extra passed to the intent. So we only have the assertion for component, which is mandatory for all intents. Nothing else.
>    private fun createProfileChooserActivityIntent(): Intent {
>    return ProfileChooserActivity
>    .createProfileChooserActivity(ApplicationProvider.getApplicationContext())
>    }
> 7. When testing the intents, you need to test for what is there. All intents must have a component to which is the Class to be opened by that intent, so all tests must have the `hasComponent()` assertion.
> 8. Some activities will not have other extras but will have been decorated with profileId, so in addition to `hasComponent`, there will be an assertion to check if intended has a profileId (we discussed how to do this in a different thread).
> 9. Some activities have only a proto extra in the bundle, so your test must contain an hasProtoextra() assertion that expects an argument, and you will create this as per the example in my last comment.
> 10. Other times, an intent will have both the extra args, and will also have decorate with profile Id. We expect to see 3 assertions for component, protoextra and profileId key.
> 11. It is your responsibility to countercheck that the intent in the test file matches the corresponding production class intent. You must make note of the items that are in the intent, and assert them in your test.
> 12. Most of the tests already have the correct assertions, and the major changes in them will be to replace hasextra/hasextrawithKey with hasProtoextra. Or add the check for profileId key. Or both.
> 
> Does this help make things clearer?

Referencing 9th point
I'm asking about the values for the arguments which I assert using hasProtoaExtra() for adminauthacitvity, I'm not getting what values to give, 
i have no idea what arguments values it expects, because there is not intent creation for Adminauthacitvity in this test file.

In this test file only profilechooserfragment intent creation is there.

Reply:
The activity being created here is profile chooser. Please see line 453/454 and refer to point number 2 again.

Reply:
![IMG_20231221_174116](https://github.com/oppia/oppia-android/assets/59600948/406ba265-8bb7-4fc7-96d6-9e0360513541)

Reply:
> The activity being created here is profile chooser. Please see line 453/454 and refer to point number 2 again.

So you mean here I'm supposed to assert profileChooserActivity extras ? if yes, my doubt is why asserting AdminAuthActivity extras in old test ?

 ```
intended(hasComponent(AdminAuthActivity::class.java.name))
      intended(hasExtra(ADMIN_AUTH_ENUM_EXTRA_KEY, AdminAuthEnum.PROFILE_ADMIN_CONTROLS.value))
```

Reply:
![Screenshot_2023-12-22-06-57-08-786_org oppia android](https://github.com/oppia/oppia-android/assets/59600948/f694e594-f95e-41d6-9b05-5f91b1678bb6)


This is the profile chooser, and at the bottom, you can open admin controls. 

The intent from splash activity(I believe) to the chooser contains that key for the admin controls bit.

Why don't you follow that flow in the code and let me know what you come up with?

Reply:
> ![Screenshot_2023-12-22-06-57-08-786_org oppia android](https://private-user-images.githubusercontent.com/59600948/292370055-f694e594-f95e-41d6-9b05-5f91b1678bb6.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMyMTg3MDgsIm5iZiI6MTcwMzIxODQwOCwicGF0aCI6Ii81OTYwMDk0OC8yOTIzNzAwNTUtZjY5NGU1OTQtZjk1ZS00MWQ2LTliMDUtNWY5MWIxNjc4YmI2LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjIyVDA0MTMyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiOTRjZDg5NmMwNWI4ZjNmNzU5YmVmOTA0NjQ0ODU3MjM4YWEwNjg4ZGYyNWQyNDg0YmZmZGI5YWI1ZGYxZjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.lK16qVDgDE1Oz4lKprn3yGWBqNv-pCqJt7oiULwdpAs)
> 
> This is the profile chooser, and at the bottom, you can open admin controls.
> 
> The intent from splash activity(I believe) to the chooser contains that key for the admin controls bit.
> 
> Why don't you follow that flow in the code and let me know what you come up with?

I get key but how i check the value for the extras ? and here test is for checking adminAuthActivity opens or not when user clicks on administrator controls(on the bottom of profile chooser screen)

Reply:
> > ![Screenshot_2023-12-22-06-57-08-786_org oppia android](https://private-user-images.githubusercontent.com/59600948/292370055-f694e594-f95e-41d6-9b05-5f91b1678bb6.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMyMTg3MDgsIm5iZiI6MTcwMzIxODQwOCwicGF0aCI6Ii81OTYwMDk0OC8yOTIzNzAwNTUtZjY5NGU1OTQtZjk1ZS00MWQ2LTliMDUtNWY5MWIxNjc4YmI2LmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjIyVDA0MTMyOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiOTRjZDg5NmMwNWI4ZjNmNzU5YmVmOTA0NjQ0ODU3MjM4YWEwNjg4ZGYyNWQyNDg0YmZmZGI5YWI1ZGYxZjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.lK16qVDgDE1Oz4lKprn3yGWBqNv-pCqJt7oiULwdpAs)
> > This is the profile chooser, and at the bottom, you can open admin controls.
> > The intent from splash activity(I believe) to the chooser contains that key for the admin controls bit.
> > Why don't you follow that flow in the code and let me know what you come up with?
> 
> I get key but how i check the value for the extras ? and here test is for checking adminAuthActivity opens or not when user clicks on administrator controls(on the bottom of profile chooser screen). Do you mind writing code for the same test so that i get clarity please?

Hi @adhiamboperes, can you please follow up?

Reply:
BTW this test is not about  asserting intent extras. We didn't created or mocked AdminAuthActivity so i don't know the exact values to pass.

Reply:
```
@Test
  fun testAdminAuthActivity_intent_extras_passed() {
    activityTestRule.launchActivity(
      AdminAuthActivity.createAdminAuthActivityIntent(
        context = context,
        adminPin = "12345",
        profileId = internalProfileId,
        colorRgb = -10710042,
        adminPinEnum = AdminAuthEnum.PROFILE_ADD_PROFILE.value
      )
    )

    val args = AdminAuthActivityArguments.newBuilder().apply {
      this.adminPin = "12345"
      this.internalProfileId = this@AdminAuthActivityTest.internalProfileId
      this.colorRgb = -10710042
      this.adminPinEnum = AdminAuthEnum.PROFILE_ADD_PROFILE.value
    }.build()

    intended(hasProtoExtra(AdminAuthActivity.ADMIN_AUTH_ACTIVITY_ARGUMENTS_KEY, args))

  }
```
Like here i wrote test for `AdminAuthActivity's `  intent extras, here i launched AdminAuthActivity  with intent extras so i know what values to pass while testing.

Reply:
Like here in old test  ` intended(hasExtra(ADMIN_AUTH_ENUM_EXTRA_KEY, AdminAuthEnum.PROFILE_ADMIN_CONTROLS.value))` only testing for adminAuthEnum beacuse thats what only could be predicted because it being passed by default when creating intent in AdminAuthActivity, Other values cannot be predicted, either we have to create helper test activity to pass predefined intent extras so that later in test we can use them to test intent extras or we can go with `hasExtraWithKey `.

Reply:
This is what I have understood so far now.

Reply:
okay